[JIT] enable prim::isinstance to handle subtyping of DictType values and keys

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47299 [JIT] enable prim::isinstance to handle subtyping of DictType values and keys**

Extended prim::isinstance to let DictTypes subtype other DictTypes if their keys and values subtype.  

Solves issue where `torch.jit.isinstance(x, Dict[int, torch.Tensor])` would return false for any `x` which had a value type of a different tensor type.

Differential Revision: [D24709998](https://our.internmc.facebook.com/intern/diff/D24709998)

Commit:
[JIT] enable prim::isinstance to handle subtyping of DictType values and keys

[ghstack-poisoned]

Commit:
Update on "[JIT] enable prim::isinstance to handle subtyping of DictType values and keys"




[ghstack-poisoned]

Commit:
Update on "[JIT] enable prim::isinstance to handle subtyping of DictType values and keys"


Differential Revision: [D24709998](https://our.internmc.facebook.com/intern/diff/D24709998)

[ghstack-poisoned]

SplitInfinity commented on aten/src/ATen/test/type_test.cpp:
```diff
@@ -205,4 +206,45 @@ TEST(TypeEquality, NamedTupleEquality) {
+  c10::DictTypePtr dictStrTensorInt = DictType::create(
+      StringType::get(), tensorIntPointer);
+
+  EXPECT_TRUE(dictStrTensorFloat->isSubtypeOfExt(dictStrTensorPlain, nullptr));
+  EXPECT_TRUE(dictStrTensorInt->isSubtypeOfExt(dictStrTensorPlain, nullptr));
+  EXPECT_FALSE(dictStrTensorInt->isSubtypeOfExt(dictIntTensorPlain, nullptr));
+  EXPECT_FALSE(dictStrTensorInt->isSubtypeOfExt(dictStrTensorFloat, nullptr));
+  EXPECT_FALSE(dictStrTensorPlain->isSubtypeOfExt(dictStrTensorInt, nullptr));
```
Consider using `isSubtypeOf` since you aren't using the `why_not` argument.

SplitInfinity commented on test/jit/test_isinstance.py:
```diff
@@ -51,6 +51,14 @@ def list_str_test(x: Any):
         x = ["1", "2", "3"]
         self.checkScript(list_str_test, (x,))
 
+    def test_list_tensor(self):
```
Do we need a `test_list_tuple` that tests `Tensor` subtyping also?

Lilyjjo replied:
Yeah, added it.

SplitInfinity reviewed (commented):
Nice catch! Just a few questions but otherwise looks good to go.

Lilyjjo commented on aten/src/ATen/test/type_test.cpp:
```diff
@@ -205,4 +206,45 @@ TEST(TypeEquality, NamedTupleEquality) {
+  c10::DictTypePtr dictStrTensorInt = DictType::create(
+      StringType::get(), tensorIntPointer);
+
+  EXPECT_TRUE(dictStrTensorFloat->isSubtypeOfExt(dictStrTensorPlain, nullptr));
+  EXPECT_TRUE(dictStrTensorInt->isSubtypeOfExt(dictStrTensorPlain, nullptr));
```
I added a test case that uses TensorTypes for the keys and for the values to cover that fun case 👍

Commit:
Update on "[JIT] enable prim::isinstance to handle subtyping of DictType values and keys"


Extended prim::isinstance to let DictTypes subtype other DictTypes if their keys and values subtype.  

Solves issue where `torch.jit.isinstance(x, Dict[int, torch.Tensor])` would return false for any `x` which had a value type of a different tensor type.

Differential Revision: [D24709998](https://our.internmc.facebook.com/intern/diff/D24709998)

[ghstack-poisoned]

eellison reviewed (changes requested):
I think this will be fixed by https://github.com/pytorch/pytorch/pull/46130. 
I don't think we want to check the change subtyping rules, that breaks type safety: https://hhvm.com/blog/9215/covariance-contravariance-and-super-type-constraints
```
@torch.jit.script
def foo():
    x: Dict[int, Optional[int]] = {1: None}
    assert torch.jit.isinstance(x, Dict[int, int]])
    # this does not return an int, not type safe
    return x[1] is not None
```

Lilyjjo commented:
@eellison Do you know why `isinstance(x, List[torch.Tensor])` already works for lists of Tensors? I was trying to figure out how the logic for Lists checks/creation differ from Dicts but I wasn't able to figure it out

eellison commented:
@Lilyjjo i think you were running into this problem:
> For Any, we infer the type of the ivalue to set the ivalue's type tag. When we saw a Tensor, we would use a specialized Tensor type, so when Dict[str, Tensor] was passed in as any Any arg it would be inferred as Dict[str, Float(2, 2, 2, 2)] which breaks runtime isinstance checking

I would be surprised if the List has different behavior here wrt/ passing it in as Any, but if there it's something in pybind_utils.h
